sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -153,7 +153,9 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
       RewritePredicateSubquery,
       ColumnPruning,
       CollapseProject,
-      RemoveRedundantProject)
+      RemoveRedundantProject) :+
+    Batch("UpdateAttributeReferences", Once,
+      UpdateNullabilityInAttributeReferences)
```

Comment:
I noticed that this rule is very similar to `FixNullability`. Do we have to keep both of them? cc @maropu @gatorsmile

Reply:
`UpdateNullabilityInAttributeReferences ` is an optimizer rule that was added after we introduced the analyzer rule `FixNullability`. Do we have any end-to-end test case for showing the benefit of this optimizer rule?

Reply:
But these 2 rules are almost same, except that `FixNullability` skips resolved plan.

Reply:
Yea, I see... But, I have no idea about the way to resolve the two issue (the correctness issue and the optimization issue) simultaneously in a single place. I think there is one greedy & temporary solution that merges the current two rule into one, and then put it in the two places: the analyzer and the optimizer.